engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/LazyDiagnosticStorage.scala

```diff
@@ -0,0 +1,32 @@
+package org.enso.compiler.core.ir
+
+import org.enso.compiler.core.IR
+
+trait LazyDiagnosticStorage { self: IR =>
+
+  private[this] var _diagnostics: DiagnosticStorage = _
+
+  protected def diagnostics_=(diagnostics: DiagnosticStorage): Unit = {
+    assert(_diagnostics eq null)
+    _diagnostics = diagnostics
+  }
+
+  override def diagnostics: DiagnosticStorage = {
```

Comment:
I am not really sure what `_diagnostics: DiagnosticStorage = _`, but if I parse the ideas in this file, I would really prefer to get only:
```java
interface IR  {
  DiagnosticsStorage diagnostics(boolean canReturnNull);
}
```
That's all we need to defer allocation of `DiagnosticsStorage` until needed, in my opinion. The boilerplate of `diagnosticsList`, `getDiagnostics`, `diagnosticsCopy` can be moved elsewhere or preferably disappear.

Reply:
```
_diagnostics: DiagnosticStorage = _
```
represents the actual allocated `DiagnosticStorage` and `_` means null in this context. You call it `d` he calls it `_diagnostics` which is more in line with Scala's way of defining custom setters/getters. I think this is fine.
Also you would have to `canReturnNull` all over the place and I don't think this is any better.

Reply:
With this approach
```scala
ir.diagnostics(canReturnNull = true).copy()
```
can throw NPE, that's why there is a to DRY the null checking logic
```scala
ir.diagnosticsCopy()
```

Reply:
I think I can remove the extra methods from the `IR` interface and make them more implementation-specific.

Reply:
> With this approach
> 
> ```scala
> ir.diagnostics(canReturnNull = true).copy()
> ```
> 
> can throw NPE, that's why there is a to DRY the null checking logic

In order to be DRY for internal stuff you duplicate a single _publicly visible_ method four times! That's not DRY! To be DRY you'd write a helper method somewhere (hidden enough from regular users of the `IR`) and used it in all the internal implementations to avoid NPE:
```java
public static DiagnosticsStorage copyOrNull(DiagnosticsStorage sOrNull) {
  return sOrNull == null ? null : sOrNull.copy();
}
```
and then called it as `DiagnosticsStorage.copyOrNull(ir.diagnostics(true))`.

> You call it d he calls it

I don't care about any of the proposed names. I just believe four methods related to diagnostics in `IR` is 75% too much. I believe single `DiagnosticsStorage diagnotics(boolean canReturnNull)` would be enough to satisfy all use-cases and keep [conceptual surface](http://wiki.apidesign.org/wiki/Conceptual_surface) (some may call it DRY) smaller.